### ISSUE TITLE
[Suggestion] useFullscreen

### ISSUE THREAD

**Issue by sszczep** [imbhargav5/rooks#1235]
Hey,

There have been a lot of issues with useFullscreen hook and I noticed that there is a rewritten version on one of the dev branches. I'm strongly against removing polyfill as stated in #579. The suggested polyfill in documentation ([fullscreen-polyfill](https://www.npmjs.com/package/fullscreen-polyfill)) won't work with Next.JS and other SSR. I'm also concerned about early return from hook (you even disabled the eslint rule). As I needed a solution to all above problems ASAP, I've created my own. 

[codesandbox](https://codesandbox.io/s/usefullscreen-8k389n?file=/src/useFullscreen.ts)
[demo](https://8k389n.csb.app) (fullscreen is not allowed in iframe on codesandbox)

I think it solves all the problems, is type-safe and also SSR friendly. 

Please let me know if you would be interested in including it in next major release or you'd rather stick with your implementation. Of course if it needs changes also let me know.

**Comment by imbhargav5** [comment]
@sszczep Thanks for this! useFullscreen will be rewritten for sure for v7. It has a poor code quality and is buggy and the plan is to rewrite the whole thing in the next 2 weeks and I would greatly appreciate a PR!

Re: the polyfill. Is there a different polyfill that we can use that works with Next.js and other SSR libs? I really am not keen to increase bundle size by adding polyfills and if there is a way for users to choose their polyfill I want to leave it to them.

**Comment by sszczep** [comment]
I don't think external polyfill is feasible, couldn't find any. 

I think the only solution is to include it using script tag (I really don't like that, what if there is a race condition and react mounts before the script executes? Seems error prone and requiring exhaustive testing) or by using next dynamic import (I don't like that either, it's platform specific and other renderers might not have such functionality. Also might still be impacted by race condition).

I stand by providing polyfill and not requiring users to manage it on their own. Our second option is to provide a Fullscreen API as a dependency. We would still provide our own polyfill, but allow users to use their own. Consider following code:

```ts
export interface FullscreenAPI {
  requestFullscreen: Element['requestFullscreen'];
  exitFullscreen: Document['exitFullscreen'];
  // ...
}

// Polyfilled Fullscreen API
const fullscreenAPI: FullscreenAPI = {
  requestFullscreen: (options?: FullscreenOptions | undefined) => Promise.resolve(),
  exitFullscreen: () => Promise.resolve(),
  // ...
};

// Returns hook with provided API
export function createUseFullscreen(fullscreenApi: FullscreenAPI) {
  // ...
}

export default createUseFullscreen(fullscreenAPI);
```

Usage:
```ts
// With default polyfill:
const { enable, disable } = useFullscreen();

// With user's polyfill:
const { enable, disable } = createUseFullscreen(myCustomImplementation)();
```

It's API agnostic, easily extensible and allows for tree-shaking - you were worried about bundle size, our polyfill will not be included in the bundle if users choose not to. 

Please tell me what you think.

**Comment by imbhargav5** [comment]
@sszczep Got it! Let's keep the polyfill and let's not bother with exporting another way for users to add their own polyfill. Your code looks great! Would you like to submit a PR? 🔥🔥

**Comment by sszczep** [comment]
I will make further adjustments (better error handling etc.) and create a PR. We should talk about the API though. To be honest I would return only isFullscreenAvailable, isFullscreenEnabled, enableFullscreen, disableFullscreen, toggleFullscreen. 

I wouldn't return the fullscreenElement. It requires one more state in hook and I can't find any use case for it. 

onChange and onError callbacks are also not something strictly necessary, but I would leave them for completeness sake (eg. browser might not support promises in Fullscreen API and one might need a more sophisticated error handling). 

Tell me what you think and I'll get on it.

**Comment by imbhargav5** [comment]
> onChange and onError callbacks are also not something strictly necessary, but I would leave them for completeness sake (eg. browser might not support promises in Fullscreen API and one might need a more sophisticated error handling).

Great callout. Let's do that. 

> I wouldn't return the fullscreenElement. It requires one more state in hook and I can't find any use case for it.

Fine by me. I don't think there is much of a reason to expose it, although it's harmless and doesn't really cause any extra rerender.

**Comment by sszczep** [comment]
> > I wouldn't return the fullscreenElement. It requires one more state in hook and I can't find any use case for it.
> 
> 
> 
> Fine by me. I don't think there is much of a reason to expose it, although it's harmless and doesn't really cause any extra rerender.

I guess you're right. It's not a big issue to add it. Maybe someone would want to check if specific element is fullscreened.

I will PR once ready

**Comment by imbhargav5** [comment]
Let's go! @sszczep 🔥🔥

**Comment by sszczep** [comment]
Solved in #1241
